Add basic pages block

tellthemachines opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->

Closes #23689 .

Adds a dynamic Pages block that updates with new pages and can be added inside a Navigation block or used on its own.

**Update: this is now ready for review.**

~~I am using `wp_list_pages` to output the list of pages, and have extended `Walker_Page` to add the chevron to items containing children. Apart from that, I have kept the default output classnames. If we'd rather be more consistent with other block classnames, I can further extend the walker to change the `li` and submenu `ul` classes, but can't do much more than that, e.g. the `a` elements aren't changeable. Cc @shaunandrews @jasmussen~~

**Upadate 2**: This is now querying the pages and building the nested list manually; classnames are semantically closer to the Navigation block but not context-aware. See [here](https://github.com/WordPress/gutenberg/pull/28265#issuecomment-772207536) for more details.

The editor view is using `ServerSideRender` because there's a notable performance benefit relative to rendering REST API data with JS. This means there's an extra `div` wrapper around the block contents in the editor, which there doesn't seem to be any way to stop `ServerSideRender` from outputting.

When the block is nested inside a Nav block, all settings are inherited except for custom colors, because there's a problem with how inline styles are set for Nav block children that needs to be solved separately.

~~I haven't added any fixtures or other tests yet, will do that once we determine the markup is all correct 😅~~ Tests added/fixed.

Speaking of which, this block rendering a `ul` container is inconsistent with the current Nav block markup which expects all its children to be `li` elements. I'm expecting that to be fixed by #24644 / #24364. The Pages block shouldn't be contained in a `li` as it can be used separately from the Nav block.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

tellthemachines committed:
Add basic pages block

jasmussen commented:
Thanks so much for working on this!

shaunandrews commented:
Very cool to see this in the works! Here's what I see when the Pages block lives alone:

<img width="720" alt="image" src="https://user-images.githubusercontent.com/191598/104955138-b2d26a00-5997-11eb-8304-10fcb21efd73.png">

And here's how it looks when placed inside the Navigation block:

<img width="252" alt="image" src="https://user-images.githubusercontent.com/191598/104955158-be259580-5997-11eb-85c7-0acae7344263.png">

tellthemachines committed:
Add dropdowns and indicators

tellthemachines commented:
I've got most of the styles for displaying submenus as dropdowns in the nav block, but dropdown background colour isn't working yet because Pages isn't inheriting the styles from the parent yet. 

Submenu icons is also not wired to the parent yet - but I'm not sure it makes sense to have it as a setting in the standalone Pages block, unless we provide the option to use dropdowns there, too. The problem is I can't think of a way to render the icons in Pages unless that exists as a setting 🤔

jasmussen commented:
> Submenu icons is also not wired to the parent yet - but I'm not sure it makes sense to have it as a setting in the standalone Pages block, unless we provide the option to use dropdowns there, too. The problem is I can't think of a way to render the icons in Pages unless that exists as a setting 🤔

Can the pages block inherit whatever property is set on the Navigation Container? I.e. if the pages block is inside the navigation block, and the navigation block has the option toggled on, so does the Pages block?

tellthemachines committed:
Inherit settings from parent Nav block

jasmussen commented:
This is so good. So so good. Thank you for working on this. Here's what I see:

<img width="879" alt="Screenshot 2021-01-21 at 08 47 04" src="https://user-images.githubusercontent.com/1204802/105319558-56ed2880-5bc5-11eb-873a-f817b276d837.png">

That's the Pages block inside navigation. I then proceeded to create a new page, "Journal", and sure enough, it shows up automatically when I reload:

<img width="723" alt="Screenshot 2021-01-21 at 08 47 31" src="https://user-images.githubusercontent.com/1204802/105319590-653b4480-5bc5-11eb-913a-2320ee4ca663.png">

This is the promise of the block itself, and it feels totally validated by this. 

I need to do some deeper testing, specifically with regards to how subpages work. But just as an initial test, this is very promising. 

Some small things:

<img width="335" alt="Screenshot 2021-01-21 at 08 46 24" src="https://user-images.githubusercontent.com/1204802/105319683-813ee600-5bc5-11eb-9a73-655272f6c15d.png">

Can we make it so when you press "Add all pages", it adds just this one block, as opposed to all the individual menu links?

<img width="700" alt="Screenshot 2021-01-21 at 08 46 54" src="https://user-images.githubusercontent.com/1204802/105319672-7e43f580-5bc5-11eb-9a8b-dd41769c79eb.png">

We should probably rename this one from "Pages" to something else. "Page Listing" was suggested, that might be worth a try.

I'd also like to provide you with a new icon for this one, I shall return. 

Finally, something is making the markup and visual appearance differ from individual menu items, compared to items from that page listing. I have a black underline in my theme that shows up for the page listing (see above), but it's gone for individual items:

<img width="779" alt="Screenshot 2021-01-21 at 08 46 28" src="https://user-images.githubusercontent.com/1204802/105319939-d975e800-5bc5-11eb-9c15-b1658de8841f.png">

I'd like to dive in and see what that might be and how I can potentially fix it. It's probably a separate thing regardless, but thought I should mention it.

The vertical menu still shows up as horizontal:

<img width="752" alt="Screenshot 2021-01-21 at 08 52 25" src="https://user-images.githubusercontent.com/1204802/105320119-14781b80-5bc6-11eb-888f-6e1c06668df3.png">

I can also click on the menu items:

<img width="926" alt="Screenshot 2021-01-21 at 08 52 34" src="https://user-images.githubusercontent.com/1204802/105320133-193ccf80-5bc6-11eb-937a-e6b8bc7926c1.png">

We can probably make the menu items inactive using CSS, `pointer-events: none;` — that's also necessary for the Page Listing block to be selectable at all.

I was able to select the page listing block using the block navigator:

<img width="789" alt="Screenshot 2021-01-21 at 08 54 15" src="https://user-images.githubusercontent.com/1204802/105320290-4a1d0480-5bc6-11eb-9b00-53a4f7eef012.png">

How hard would it be to add a button in the page listing block toolbar that says "Convert to individual menu items"? (Or a shorter phrase if we can think of one) 

Thanks again for your work 🏅

talldan commented:
Working really nicely so far.

I'd worry that the heavy reliance on server side rendering and the walker might end up being a limiting factor in future changes, like if we want to change the markup. 

The alternative (making it work more like the query block does) would be pretty complicated though, so might be best to go for the simple solution first and see how well it works (which I think is pretty successful so far).

tellthemachines committed:
Add pages button adds the block and styling improvements

tellthemachines commented:
> Can we make it so when you press "Add all pages", it adds just this one block, as opposed to all the individual menu links?

Done ✅ 

> We should probably rename this one from "Pages" to something else. "Page Listing" was suggested, that might be worth a try.

Could it be "Page list" instead? "Listing" sounds more like a single listed item, so may be confusing.

> I have a black underline in my theme that shows up for the page listing (see above), but it's gone for individual items:

In the editor, it's the default browser `a` styles being applied, and that doesn't happen for individual items because they aren't `a` but contenteditable `div`s. In the front end, it's generic theme styling for `a` (guessing that based on TT1), that is overridden for Navigation link blocks specifically. I guess we'll have to wait for a theme update to fix it once this block is merged.

> The vertical menu still shows up as horizontal
> I can also click on the menu items

Fixed, and fixed! 

> How hard would it be to add a button in the page listing block toolbar that says "Convert to individual menu items"?

I thought initially that it would be simple to transform a Pages block into a Links block once #24644 is done, but given that Pages should work as a standalone block, would we want that functionality to also apply when it isn't part of a Navigation? We don't have any way of conditionally rendering controls depending on the parent block (I think - because blocks aren't supposed to be aware of their parents) so not sure how we'd go about making it Navigation specific.

> The alternative (making it work more like the query block does) would be pretty complicated though

I'm more worried both about the performance impact (which is hard to tell now because query only renders 10 items at a time - unless I'm missing something) and the duplication of markup (query loop has its `ul` and `li` structure repeated in both `edit`  and `index.php`, which is not ideal, especially when we have more complex markup with nested lists).

> TT1-Blocks has a couple of issues (pages always appear vertically, and submenus don't open

Hmm, I can't reproduce that by adding a Nav block with a Pages block inside the header in the site editor 🤔

jasmussen commented:
> Could it be "Page list" instead?

Go for it!

> In the editor, it's the default browser a styles being applied, and that doesn't happen for individual items because they aren't a but contenteditable divs. In the front end, it's generic theme styling for a (guessing that based on TT1), that is overridden for Navigation link blocks specifically. I guess we'll have to wait for a theme update to fix it once this block is merged.

Thanks for the investigative work. Seems like it would be nice if we were able to change those contenteditable divs to contenteditable `a` tags. Not even sure if that's possible, and would need to be a separate thing regardless. Otherwise, fixable easily in an editor style. Thank you.

> I thought initially that it would be simple to transform a Pages block into a Links block once #24644 is done, but given that Pages should work as a standalone block, would we want that functionality to also apply when it isn't part of a Navigation? We don't have any way of conditionally rendering controls depending on the parent block (I think - because blocks aren't supposed to be aware of their parents) so not sure how we'd go about making it Navigation specific.

Definitely best to try and avoid making anything Navigation specific if we can. But "navigation aware" might not be the end of the world if it's done in a clean way.

Specifically, I'm not sure that this Pages (or Page List) block _has_ to work outside of the navigation block. I think that's a really nice benefit.

And yes, I think it would be totally fine to have that functionality also work when used on its own. I compare it to a gallery of 5 images, which you can transform to 5 individual image blocks. (I realize here the transformation from Page List to individual Paragraphs with links inside them would be an irreversible transformation, but I don't see that necessarily as a problem).

talldan commented:
> I thought initially that it would be simple to transform a Pages block into a Links block once #24644 is done, but given that Pages should work as a standalone block, would we want that functionality to also apply when it isn't part of a Navigation? We don't have any way of conditionally rendering controls depending on the parent block (I think - because blocks aren't supposed to be aware of their parents) so not sure how we'd go about making it Navigation specific.

This might not be an issue, I don't think the transform system will show a transform to a block that can't be inserted.

Worth double-checking, but an example is that you can't transform a Nav Link to a Group because Group isn't allowed in Navigation.

jasmussen commented:
I am so excited for this PR to land, it really improves things. 

Instead of the "page" icon you're using for the new block, I created a new icon you can add as "pages":

<img width="393" alt="Screenshot 2021-01-25 at 12 39 39" src="https://user-images.githubusercontent.com/1204802/105701542-bcfde680-5f0a-11eb-9780-7c759cdbe33b.png">

Here's the SVG:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7 13.8h6v-1.5H7v1.5zM18 16V4c0-1.1-.9-2-2-2H6c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h10c1.1 0 2-.9 2-2zM5.5 16V4c0-.3.2-.5.5-.5h10c.3 0 .5.2.5.5v12c0 .3-.2.5-.5.5H6c-.3 0-.5-.2-.5-.5zM7 10.5h8V9H7v1.5zm0-3.3h8V5.8H7v1.4zM20.2 6v13c0 .7-.6 1.2-1.2 1.2H8v1.5h11c1.5 0 2.7-1.2 2.7-2.8V6h-1.5z" /></svg>
```

Let me know if you'd like me to push directly to your branch to add this.

jasmussen commented:
This one fixes #16635.

tellthemachines committed:
Change name to page list

tellthemachines committed:
Add pages icon

tellthemachines commented:
Ok, I think I've addressed all the feedback, so if everything is now OK I'll move onto adding some tests for this block. 

The transform from Page List to Links is dependent on #24644, so let's do it as a separate PR later.

jasmussen commented:
So excited for this one, thanks again. I'm happy to push [the icon change](https://github.com/WordPress/gutenberg/pull/28265#issuecomment-766758332) separately.

A quick question just to be sure — you mentioned how on the frontend and the page list, the links are _links_ (`a`), and custom menu items have contenteditable divs. Not necessarily as part of this branch, but can we change those divs to contentedtiable `a`'s as well? From a very quick codepen sanity check [it doesn't seem impossible](https://codepen.io/joen/pen/KKgOrGK), but I'm very probably missing something.

The consistency in markup between menu items would really simplify some of the CSS!

tellthemachines commented:
> I'm happy to push the icon change separately.

Whoops, I made the change and then forgot to push it 😳 

> Not necessarily as part of this branch, but can we change those divs to contentedtiable a's as well?

Let's try that as a separate task! I can't think offhand of reasons not to do it, but might be missing something.

tellthemachines committed:
Add page list fixture

tellthemachines committed:
Update nav screen snapshot

tellthemachines commented:
I've added the tests, so this is ready for a final review now!

jasmussen commented:
> Let's try that as a separate task! I can't think offhand of reasons not to do it, but might be missing something.

Awesome. Want me to ticket it?

I'm seeing jump when the page list loads:

<img width="595" alt="Screenshot 2021-01-28 at 12 03 27" src="https://user-images.githubusercontent.com/1204802/106130275-70f8af00-6161-11eb-9362-a561c82368a8.png">

You can fix that with this CSS:

```
.wp-block-navigation .components-placeholder {
	min-height: 0;
	padding: 0;

	.components-spinner {
		margin-top: 0;
	}
}
```

Then it looks like this:

<img width="296" alt="Screenshot 2021-01-28 at 12 08 33" src="https://user-images.githubusercontent.com/1204802/106130425-9ab1d600-6161-11eb-819e-e27e9c7ac29b.png">

---

I noticed submenus have no background:

<img width="539" alt="Screenshot 2021-01-28 at 12 15 55" src="https://user-images.githubusercontent.com/1204802/106131213-b073cb00-6162-11eb-984d-a379f3715bc4.png">

This is because the markup for the child blocks is this:

```
<ul class="children">
	<li class="page_item page-item-263"><a href="http://localhost:8888/?page_id=263">This is a subpage</a></li>
	<li class="page_item page-item-265"><a href="http://localhost:8888/?page_id=265">This is another subpage</a></li>
</ul>
```

This is compared to the markup of "classic" navigation submenus:

```
<ul class="wp-block-navigation__container">
   <li class="wp-block-navigation-link"><a class="wp-block-navigation-link__content" href="http://localhost:8888/?page_id=265"><span class="wp-block-navigation-link__label">This is another subpage</span></a></li>
   <li class="wp-block-navigation-link"><a class="wp-block-navigation-link__content" href="http://localhost:8888/?page_id=263"><span class="wp-block-navigation-link__label">This is a subpage</span></a></li>
</ul>
```

I know there's a goal to make this Page List block work outside of the navigation block. Can it be aware of its parent, and show the navigation-block appropriate CSS classes when it's used there? 

---

Finally, I'm really missing that feature to convert a Page List block into menu items. Whether that's a transformation, or a button in the toolbar, "Convert to menu items", it's important that we land this feature. Doesn't necessarily have to be part of this PR specifically, but we should have some confidence that it's possible to do, so we can follow up with it.

Those were the three things I found. But the primary take-away: this is feeling so good, and it even feels sort of obvious and predictable. I think it will solve a bunch of headaches, thank you again for working on this!

draganescu commented:
Adding the new block in the navigation block produced this result on Edge:

Front end:

![image](https://user-images.githubusercontent.com/107534/106131097-e7e67580-616a-11eb-822c-e27e6ca63d2e.png)

Back end:

![image](https://user-images.githubusercontent.com/107534/106131172-0482ad80-616b-11eb-9acf-85095f66a889.png)

Also clicking on the page items navigates to the page, which is different from the behavior of the other links in paragraphs or navigation menus, that don't navigate.

With children the styling is still different, is it intentional?

![image](https://user-images.githubusercontent.com/107534/106131991-0731d280-616c-11eb-9b65-2e27e96d2401.png)

From the PR's description it looks like this is expected since I guess the differences in markup make this block appear different.  Will a solution to #24644 / #24364. also fix these visual inconsistencies?

talldan commented on packages/block-library/src/page-list/edit.js:
```diff
@@ -0,0 +1,43 @@
+			clientId,
+			'core/navigation'
+		)[ 0 ];
+		return getBlockAttributes( parentBlock );
```
I think we should be able to use block context instead of selecting from the navigation block's attributes via the block-editor store.

The good thing is it's a bit easier, it's just passed in as a prop:
https://developer.wordpress.org/block-editor/developers/block-api/block-context/#using-block-context

Looks like we should also make this change to nav link at some point as well.

talldan commented on packages/block-library/src/page-list/index.js:
```diff
@@ -0,0 +1,23 @@
+	title: _x( 'Page List', 'block title' ),
+	description: __( 'Display a list of all pages.' ),
+	icon,
+	example: {},
```
Feels weird that this is defined as an empty object, but I see other blocks are doing this too, so whatever 😄

talldan commented on packages/block-library/src/page-list/index.js:
```diff
@@ -0,0 +1,23 @@
+	icon,
+	example: {},
+	edit,
+};
```
Do we need keywords? Maybe not, I can't think of any.

talldan replied:
'Menu' could be one, given it'll be useful for building menus. Also 'Navigation'?

tellthemachines replied:
I can't think of any either 😅  We can always add them later if needed?

talldan commented on packages/block-library/src/page-list/editor.scss:
```diff
@@ -0,0 +1,11 @@
+	.wp-block-page-list {
+		background-color: inherit;
+	}
+	// Make links unclickable in the editor
```
I think the `<Disabled>` component was the previous way of doing this (see the Archives block). Not sure what the pros/cons are of the two ways.

tellthemachines replied:
Hmm, that won't work here because `Disabled` will prevent any dropdowns from opening. It works for Archives because there are no nesting levels there. It uses the same CSS to achieve the purpose though.

talldan commented on packages/e2e-tests/specs/experiments/__snapshots__/navigation-editor.test.js.snap:
```diff
@@ -2,11 +2,7 @@
-<!-- wp:navigation-link {\\"label\\":\\"About\\",\\"type\\":\\"page\\",\\"id\\":2,\\"url\\":\\"https://this/is/a/test/page/about\\"} /-->
-
-<!-- wp:navigation-link {\\"label\\":\\"Contact\\",\\"type\\":\\"page\\",\\"id\\":3,\\"url\\":\\"https://this/is/a/test/page/contact\\"} /-->
+<!-- wp:page-list /-->
```
Hmm, this test doesn't really assert much now that the snapshot doesn't show the menu structure. We don't know if page-list is actually working. I wonder what else we can do here.

tellthemachines replied:
How about we do something like `page.$$('.page-item')` and check that it returns the expected number of items?

talldan replied:
Yeah, I think that'd be good 👍

talldan reviewed (commented):
Others have done a fair amount of testing, so this review focuses mostly on code. Looks good so far.

I'm also still not sure about using the walker approach, we can always try different things once shipped if it feels like it's not working for us. I think that's one of the main reasons we should keep it behind a the GUTENBERG_PHASE flag for now (although I think it'll miss 5.7 anyway).

tellthemachines committed:
Fix placeholder and dropdown background

tellthemachines commented on packages/block-library/src/index.js:
```diff
@@ -148,6 +149,7 @@ export const __experimentalGetCoreBlocks = () => [
 	missing,
 	more,
 	nextpage,
+	pageList,
```
I'm thinking Page List can be used immediately in the same way as Archives or Categories, as we're currently missing a block replacement for the Pages widget. The work in this PR makes it usable as a standalone block; all the missing features are Navigation-specific so they're not relevant to standalone use anyway.

talldan replied:
> I'm thinking Page List can be used immediately in the same way as Archives or Categories

The flag only stops the feature being shipped in WordPress releases, it'll still be available for immediate use in the plugin. As I mentioned above, given the timing of Gutenberg 9.9 being today it doesn't really matter that much now.

tellthemachines committed:
Use context and add keywords

tellthemachines commented:
> Awesome. Want me to ticket it?

Yes please! 🍫 

I've fixed the placeholder and the submenu backgrounds. The way we're dealing with those backgrounds could be improved though, as the specificity of the editor default can override theme styles:

<img width="674" alt="Screen Shot 2021-01-29 at 4 50 02 pm" src="https://user-images.githubusercontent.com/8096000/106237079-26a32b00-6252-11eb-8924-99c525daa426.png">

If there are any plans for using CSS variables in default editor styles, we could do something like TT1 is doing and set `wp-block-navigation__container` to the default editor background-color. 

> I know there's a goal to make this Page List block work outside of the navigation block. Can it be aware of its parent, and show the navigation-block appropriate CSS classes when it's used there?

I don't _think_ so, but let me investigate further.

> Finally, I'm really missing that feature to convert a Page List block into menu items. Whether that's a transformation, or a button in the toolbar, "Convert to menu items", it's important that we land this feature. Doesn't necessarily have to be part of this PR specifically, but we should have some confidence that it's possible to do, so we can follow up with it.

This is definitely possible to do! Though we should wait until #24644 is done, because that's changing the way we deal with Links inside the Navigation block.

> Also clicking on the page items navigates to the page, which is different from the behavior of the other links in paragraphs or navigation menus, that don't navigate. 

@draganescu that was fixed a few commits ago; could you pull, rebuild and check again? 

> From the PR's description it looks like this is expected since I guess the differences in markup make this block appear different. Will a solution to #24644 / #24364. also fix these visual inconsistencies?

It might, and @jasmussen 's wish is to change the markup of Nav Link contenteditables so that would fix it too.

> I'm also still not sure about using the walker approach, we can always try different things once shipped if it feels like it's not working for us.

@talldan I'm more than happy to try something else now, but I don't know of any other way to render the pages server-side apart from `wp_list_pages`, and that means we need the walker to alter the markup. Suggestions welcome!

One thing we can do is leverage the walker further to make the classnames match block conventions more closely. That way we could potentially replace the walker at a later stage with no breaking changes.

talldan commented:
Spotted a small issue. The `wp-block-page-list` class is added a few times. Three times in the editor, and twice in the rendered post content:

Editor:
<img width="261" alt="Screenshot 2021-01-29 at 3 58 32 pm" src="https://user-images.githubusercontent.com/677833/106248120-80542700-624b-11eb-92e9-cb53ed9ff2e7.png">

Content:
<img width="353" alt="Screenshot 2021-01-29 at 4 03 46 pm" src="https://user-images.githubusercontent.com/677833/106248186-98c44180-624b-11eb-8a8b-143e86ce6fc7.png">

talldan commented:
This also fixes https://github.com/WordPress/gutenberg/issues/16635

jasmussen commented:
> > Awesome. Want me to ticket it?
> Yes please! 🍫

Ticketed here: https://github.com/WordPress/gutenberg/issues/28575

Looking at https://github.com/WordPress/gutenberg/pull/28265/commits/b7d7f92115edbb5d88c5c28f24bf1baa104f2c85, I think there's an opportunity to revisit this. Perhaps doesn't need to be this PR, because it's already big and wonderful. But in part, it doesn't work for me:

<img width="492" alt="Screenshot 2021-01-29 at 09 50 17" src="https://user-images.githubusercontent.com/1204802/106253301-96e08b00-6217-11eb-86bf-d3d51e300ded.png">

> I don't think so, but let me investigate further.

Mostly, I'm about to dive into issues of styling the CSS of the navigation block, including submenus. That means frontend and backend ideally looking identical the whole way through. This is why it would it would be so helpful if the CSS classes were the same whether a dropdown menu was inside the Page List block, or manually created.

I recall a separate conversation about creating a "dropdown menu component". Could this effort let us explore having identical CSS classes, if not markup, frontend and editor?

Thank you again, I'm so excited for this one to land!

draganescu commented:
@tellthemachines I am probably doing something wrong but making sure cache gets cleared, removed the build folder, pulled latest updates, rebuilt, on Windows in both Edge and Frirefox clicking the links in the stand alone block navigates away.

tellthemachines committed:
Fix clickability and remove extra classname.

tellthemachines committed:
Test presence of pages block items

tellthemachines commented:
> on Windows in both Edge and Frirefox clicking the links in the stand alone block navigates away.

Ha, you're right, `pointer-events: none` was scoped to the Navigation block 🤦 

Fixed that and the duplicate classnames (there's still a duplicate between the `div` and the `ul` in the editor, but I don't think that can be helped due to how `ServerSideRender` works)

shaunandrews commented:
This is really taking shape. Nice work.

--

<img width="569" alt="image" src="https://user-images.githubusercontent.com/191598/106506194-6149d500-6497-11eb-80a0-6a3802b406eb.png">

When I added the Page List block inside a Navigation block, some random extra space appears. This might be a theme (2021) issue, maybe?

--

<img width="368" alt="image" src="https://user-images.githubusercontent.com/191598/106506263-7aeb1c80-6497-11eb-839f-36ad599119e7.png">

When the Page List block is inside a Navigation block the more menu has an empty space that doesn't need to be there.

--

Is adding a "Transform to Links" option out of scope for this PR? It'd be handy to be able to break up the Page List block into individual Link blocks so I could customize a label or remove an item.

--

Similar, can we add a "Transform to Navigation" option for the Page List block, which would then wrap it in a Navigation block. Perhaps best left for another PR?

jasmussen commented:
The empty menu item may be solved by rebasing, see #28486.

tellthemachines commented:
> Mostly, I'm about to dive into issues of styling the CSS of the navigation block, including submenus. That means frontend and backend ideally looking identical the whole way through. This is why it would it would be so helpful if the CSS classes were the same whether a dropdown menu was inside the Page List block, or manually created.

That makes sense! I'm going to explore another way of rendering the markup server-side which will allow us more flexibility. Potentially we could use the "show submenu icon" context to ascertain whether the block is inside a Navigation block, though it feels slightly hacky to do so 😬 🤔 

> I recall a separate conversation about creating a "dropdown menu component". Could this effort let us explore having identical CSS classes, if not markup, frontend and editor?

Probably not in this case, because Page list doesn't have inner blocks. Its contents are not supposed to be editable, so we're generating the markup for the dropdowns server-side. That doesn't mean the classnames can't match though.

I'm thinking perhaps we could change the classname we're hooking the dropdown styles to in the Navigation block to something more generic, e.g. instead of `wp-block-navigation__container` maybe `submenu__container`? So we could use that class across different blocks, and potentially also in a future Dropdown block. Then we only need to specify styles for `.wp-block-navigation .submenu__container` and it'll also apply to any Page list that might be inside Navigation. 

(`submenu` might not be the best word here - it would be good to describe the semantics of it more than the presentation, so I'm avoiding `dropdown`. Maybe `childlist`? Which blocks would we be able to use a Dropdown block inside of?)

> Is adding a "Transform to Links" option out of scope for this PR? It'd be handy to be able to break up the Page List block into individual Link blocks so I could customize a label or remove an item.

> Similar, can we add a "Transform to Navigation" option for the Page List block, which would then wrap it in a Navigation block. Perhaps best left for another PR?

Yeah, "Transform to Links" is waiting on #24644, so best do it as a separate task. We can do a follow-up PR to add all the transforms.

jasmussen commented:
> Potentially we could use the "show submenu icon" context to ascertain whether the block is inside a Navigation block, though it feels slightly hacky to do so 😬 🤔

I believe that we will be seeing a lot more of this _blocks being aware of their containers_ in the future. For example a task block could become a task list when inside a list block. Or perhaps more imminently, a nested Image block inside a Gallery block could gain new powers of awareness of parent properties such as "Open images in new tabs". CC: @glendaviesnz in case he has input here.

tellthemachines committed:
Replace walker with manual iteration

tellthemachines commented:
Alright, I got rid of the walker and we're now iterating through the pages and building the nested list manually, so it's easier to customise the markup to our hearts' content. But we're not able to use parent block context to conditionally render markup yet: it works on the front end, but isn't recognised when using `ServerSideRender` in the editor. I think this is because the REST API `block-renderer` endpoint isn't aware of block context. 

I'll write up an issue so we can look into the best way of doing this. I'm not convinced that client-side rendering in the editor is in our best interest, as there's a noticeable performance degradation compared to server-side rendering.

In the meantime, I think we've exhausted the limits of what can be done in this PR. Final reviews welcome!

tellthemachines committed:
Fix tests

talldan commented on packages/block-library/src/page-list/index.php:
```diff
@@ -0,0 +1,216 @@
+		if ( isset( $page['children'] ) ) {
+			$css_class[] = 'has-child';
+		}
+		$markup .= '<li class="' . implode( ' ', $css_class ) . '">';
+		$markup .= '<a class="wp-block-pages-list__item__link" href="' . $page['link'] . '">' . $page['title'] . '</a>';
```
Hmm, might need to run `esc_url` over the link and potentially `wp_kses` over the title? 

Would be good to replicate the way nav link renders for safety.

tellthemachines replied:
These items aren't editable by users at any point, but better safe than sorry I guess.

I'm going to assume we allow the same HTML in post titles as in the posts themselves, until told otherwise 😅

talldan commented on packages/block-library/src/page-list/index.php:
```diff
@@ -0,0 +1,216 @@
+	}
+	$markup = '';
+	foreach ( (array) $nested_pages as $page ) {
+		$css_class = array( 'wp-block-pages-list__item' );
```
Could probably just inline `wp-block-pages-list__item` into the markup rather than making an array and then imploding it.

tellthemachines replied:
I did it that way as the `has-child` class will sometimes be added; I can just make it a string though.

talldan replied:
Oh, completely missed that. This is fine then 😄

talldan commented on packages/e2e-tests/specs/experiments/navigation-editor.test.js:
```diff
@@ -204,7 +204,13 @@ describe( 'Navigation editor', () => {
 		);
 		await addAllPagesButton.click();
 
+		await page.waitForXPath( '//a[contains(., "Sample Page")]' );
```
Previously this test mocked the pages REST API endpoint (`getPagesMocks`) to ensure it had consistent results.

While it may not be a problem right now, the test can't rely on 'Sample Page' being there, another test might remove it or change the title.

But we can't do API mocking either on the Page List block as it uses `get_pages`.

Not really sure what we can do to get the most out of this test. Maybe we should just go back to only the snapshot as the assertion, at least that checks the right block is inserted.

talldan commented on packages/block-library/src/page-list/block.json:
```diff
@@ -0,0 +1,20 @@
+	},
+	"editorStyle": "wp-block-page-list-editor",
+	"style": "wp-block-page-list"
+}
```
Should probably add a new line at the end. Other block.json files seem to have one.

talldan reviewed (commented):
Blitz those final things and I reckon this can be merged.

Great job with the refactor away from the walker. I really like this block and think it'll probably be how a lot of people build menus, given how it just automatically works.

Any thoughts on next steps? Probably filtering the pages (category/tag) would be useful.

tellthemachines commented on packages/block-library/src/page-list/index.php:
```diff
@@ -0,0 +1,216 @@
+	static $block_id = 0;
+	$block_id++;
+
+	$all_pages = get_pages();
```
Yeah these are really objects, not arrays as we understand them in JS 😅 

I'm not sure what the page order setting is supposed to do, apart from changing the order of display in `wp-admin/edit.php?post_type=page`. Trying to add all pages to a classic menu, the numbered ones are sitting between `0`s too 🤷‍♀️ 

I'd expect `get_pages` to offer a way of sorting them by order but it doesn't seem to, and I'm following the default order `get_pages` returns them in (apparently ASC).

talldan replied:
The pages widget does sort in the right order though, so probably good to replicate that, and it seems like an improvement over the old menus.

For get_pages, the `sort_column` argument should make sure the order is used for sorting. It's a shame there's no way to specify a secondary sort.

```php
get_pages( array( sort_column => 'menu_order' ) );
```

talldan replied:
I did some more investigation into this, as I became intrigued by the `hierarchical` argument of `get_pages`, which is always `true`.

It looks like if you do `get_pages( array( 'sort_column' => 'menu_order' ) )`, it by default returns all the pages in an order that takes account the nesting. The return value is a flat array, but the nested items are in the correct position. It looks like it should be possible to simplify how the nesting works in this block.

tellthemachines replied:
I'm not sure there's much simplification can be done, as we still need to nest child pages under parents to generate the proper markup, so we need the pages to be indexed by their ID, which they're not when output by `get_pages`. I can't think of an easier way of doing this, but suggestions welcome!

talldan replied:
Yeah, it is still pretty tricky. I expected `hierarchical` to build a nested array much like you've done in this PR. I think 
 to render in this format, the rendering loop would have to walk ahead to the next items to determine if they're children, and then add the nesting markup when that's the case.

Does feel a bit like an unnecessary refactoring given what's in this PR works, so maybe not worth it.

It could still be a decent optimisation to set hierarchical option to false in the pages block, as `get_pages` does a whole bunch of work that's not needed:
- https://github.com/WordPress/wordpress-develop/blob/4cfbf629ca231820cc692049be30ef588101e523/src/wp-includes/post.php#L5852-L5854
- https://github.com/WordPress/wordpress-develop/blob/4cfbf629ca231820cc692049be30ef588101e523/src/wp-includes/post.php#L5453-L5480

tellthemachines commented on packages/block-library/src/page-list/index.php:
```diff
@@ -0,0 +1,216 @@
+	if ( empty( $current_level ) ) {
+		return;
+	}
+	$nested_array = array();
```
Ha, turns out I have to reference the current item in the array by key in order to modify it. Thanks for asking or I wouldn't have bothered to find out why it wasn't working before 😁

tellthemachines committed:
Address review feedback

talldan commented on packages/block-library/src/page-list/index.php:
```diff
@@ -0,0 +1,217 @@
+	return $current_level;
+}
+
+	/**
+	 * Renders the `core/page-list` block on server.
+	 *
+	 * @param array $attributes The block attributes.
+	 * @param array $content The saved content.
+	 * @param array $block The parsed block.
+	 *
+	 * @return string Returns the page list markup.
+	 */
```
Seems like this became indented

talldan reviewed (approved):
Great work here 🎉 

Lets just see if the ordering issue can be fixed quickly before merging, but if not it can be looked at in a follow-up as an improvement.

draganescu commented on packages/block-library/src/page-list/index.php:
```diff
@@ -0,0 +1,217 @@
+
+	if ( $block->context && $block->context['showSubmenuIcon'] ) {
+		$css_classes .= ' show-submenu-icons';
+	}
```
I would move this to it's own setter like `block_core_page_add_context_classes( $context, $classes )`.

tellthemachines replied:
Would that include just this `if` statement, or the whole classname setting logic? The colours and font-size classes set above are also derived from context. 

With the submenu icons in particular, I'm hoping we can remove this class soon, and just render the submenu icons whenever they are needed. The problem is we're server side rendering the block in the editor, and `ServerSideRender` doesn't recognise context. There's some more discussion on this problem in #28684 . Anyway, I'd leave this as is for now and revisit later pending the outcome of that discussion.

tellthemachines committed:
Order menu items and de-indent comment.

jasmussen commented:
🔥

youknowriad commented on packages/block-library/src/index.js:
```diff
@@ -148,6 +149,7 @@ export const __experimentalGetCoreBlocks = () => [
 	missing,
 	more,
 	nextpage,
+	pageList,
```
So this block is stable and undependent of the navigation block, which means it's going to be included in the next WP major (5.8) regardless of the status of the Navigation block, I just want to make sure this is a conscious decision.

jasmussen replied:
@tellthemachines was the block meant to be tweaked with regards to syntax, or is it pretty much baked as is?

mtias replied:
It seems we should keep this tied to the navigation block (and its child blocks) for a bit

tellthemachines replied:
In making this an independent block, I was thinking of the Widgets editor, as we don't currently have another block that can reproduce the Pages widget functionality. Assuming we want the Widgets editor to have feature parity with the legacy widgets screen, this block will be needed. 

I'm happy with its current syntax, and the fact that it gets whatever styling properties from Navigation block context means that when used standalone, it works just like the Categories or the Archives block, that don't have any styling options.

Related question: should Categories also work as a child of Navigation? There's definitely a use case for categories in menus.

youknowriad commented on packages/block-library/src/page-list/block.json:
```diff
@@ -0,0 +1,20 @@
+		"customBackgroundColor",
+		"fontSize",
+		"customFontSize",
+		"showSubmenuIcon"
```
It feels like "context" is being abused in the navigation block, what happens if I use this block independently and I want to customize these things?

mcsf replied:
I'm only four months late to this PR, but it sounds like we should try to let Global Styles absorb some of this (cc @nosolosw, @mtias).

oandregal replied:
Looked a bit at this. It relates to the block support mechanism and parent/child relationship.

Braindump of what we have now:

- a block wants to serialize styles to its wrapper => default supports mechanism
- a block wants to serialize styles to any other node => support mechanism with skip serialization (but only for its own markup, can't affect children)
- a block wants to serialize styles to children blocks or or a children wants to take styles from the parent => block context

One thing we could explore would be to allow blocks to automatically expose some block supports via context without having to declare context support explicitely. Was that what you had in mind, Miguel?

tellthemachines replied:
I added the context in to match the Navigation Link block, so that all the links in the Navigation could be styled equally. When Page List is used as an independent block, those styling options aren't available. This seemed ok as the standalone block reproduces the functionality of the Pages widget, and similar blocks like Latest Posts also have limited styling.
As for the styling mechanism used I think we should go with the same strategy for both Navigation Link and Page List so that they work consistently inside the Navigation block, but I'm not super knowledgeable about how Global Styles work so not sure what it might look like!

mcsf replied:
> One thing we could explore would be to allow blocks to automatically expose some block supports via context without having to declare context support explicitely. Was that what you had in mind, Miguel?

Actually, I'm much less concerned about explicitly including Global Styles / Block Supports keys under `usesContext`, than I am about manually replicating some of `gutenberg_apply_colors_support` as part of `render_block_core_page_list` (see `block_core_page_list_build_css_colors`). This is where I think it'd be best to see where, if anywhere, GS/BS could improve to accommodate scenarios like Page List.

oandregal replied:
In playing with this code I've noticed the block attributes and the block context have the same shape, so we can apply the same transformations on that data to create classes and styles ― it's a matter of finding how.

A small step in making both cases more similar could be https://github.com/WordPress/gutenberg/pull/32807 (experimented only with colors to see how it feels).

mcsf commented on packages/block-library/src/page-list/index.php:
```diff
@@ -0,0 +1,217 @@
+ *
+ * @return array The nested array of pages.
+ */
+function nest_pages( $current_level, $children ) {
```
This is a pretty short name that we should be careful with as we port this to WP 5.8. At the very least, it should be `wp_nest_pages`.

youknowriad replied:
Let's try to follow-up on this during the WP 5.8 betas and clean things before the stable release.

mcsf commented on packages/block-library/src/page-list/index.php:
```diff
@@ -0,0 +1,217 @@
+	static $block_id = 0;
+	$block_id++;
+
+	$all_pages = get_pages( array( 'sort_column' => 'menu_order' ) );
```
@tellthemachines: Is there a discussion somewhere in GH about limiting this query with `number`?

tellthemachines replied:
I don't recall there being any discussion on that. I did test it with lots (several hundreds) of pages and it works ok because the block is server rendered. If we wanted to render it on the client side in the editor (I recall some talk about this at one point - possibly with @gwwar ?) it might be a different matter.

gwwar replied:
I think some reasonable limit could be set for both server/client. Any client side implementation will hit issues much sooner (eg needing to fetch all results via API+ potentially adding a ton of navigation links). 

There's should already be a limit for the page-list -> navigation links conversion, where we won't display the option if there are too many pages.
